### PR TITLE
Move li3ds-specific code in li3ds-suffixed files

### PR DESCRIPTION
**Pull request by elemoine**
I am creating this PR for discussion at this point.

This PR restructures the repo to isolate the li3ds-specific code by placing them in li3ds-suffixed files. For example this PR moves `pc_matrix.c` in `pc_matrix_li3ds.c`. And it adds `pc_patch_li3ds.c` whose content is li3ds-specific pc_patch code.

The goal is to have the non-li3ds files as close to upstream as possible, to ease merging from upstream. 

There are files that we can't split, such as `Makefile`'s, `cu_tester.c` and `pointcloud.sql.in`. For build-related files (`Makefile`'s and friends) I think it should not be a problem, as they don't change that often and conflicts should be easy to fix. For `pointcloud.sql.in`, `pointcloud.sql` and `pointcloud.out` I created a LI3DS section with the li3ds-specific bits, which should be ok as well.

@mbredif what do you think? I tried the approach we discussed last week, and other approaches as well, but they all were quite complex and required more changes to the build-related files. This one is the simplest I've come up with.

Also, one thing to note, I only handled the "transform" and "box/frustum" code that we introduced lately. There probably is other stuff that we need to move to the "li3ds" space.

### CODE REVIEW

File: pgsql/expected/pointcloud.out [hunk 1]
```diff
@@ -450,14 +437,12 @@ SELECT PC_AsText(PC_Patch(ARRAY[PC_MakePoint(3, ARRAY[-127, 45, 124.0, 4.0])]::p
 (1 row)
 
 SELECT 'pc_id1', PC_PCId(PC_Patch(PC_MakePoint(3, ARRAY[-1,-2,-3,-4])));
-WARNING:  Value -4 truncated to 0 to fit in uint16_t
```
Comment (elemoine) [inline pgsql/expected/pointcloud.out:453]: @mbredif it looks like I had to revert https://github.com/LI3DS/pointcloud/pull/60/commits/835c6999246c4ec77b40c54e88c03d9f71a4d112. Why did you add this commit? Was it required in your case?

Comment (mbredif) [reply pgsql/expected/pointcloud.out:453]: Without it, it was failing the travis CI...

Comment (elemoine) [reply pgsql/expected/pointcloud.out:453]: Now it fails with it for me, both locally and on Travis.

**Comment by mbredif** [comment]
In theory, there could be as much as 3 new modules : 
- li3ds : frustum, boxes and their transformations (no dependency)
- li3ds_pointcloud : pcpatch and pcpoint transformations, casts as frustums and boxes (dependencies : li3ds, pointcloud)
- li3ds_postgis : geom transformations, conversions from frustums and boxes to geometries (dependencies : li3ds, postgis)
- [pg_li3ds](https://github.com/LI3DS/pg_li3ds) could then depend on li3ds_pointcloud and li3ds_postgis
One question is whether it would be more favorable to start a new repository or to add these modules to the pg_li3ds repository.

In practice, your proposal is much simpler to implement. If you go this simpler route then
- pgsql/sql/pointcloud.sql additions may go to a  new pgsql/sql/li3ds.sql file (as well as the expected files)
- pgsql/pointcloud.sql.in additions could go to a new pgsql/li3ds.sql.in file (does it require a new extension ?)
- pc_*_li3ds* files and functions could be renamed to li3ds_* or li_*, at least for most of them that are not depending on pgpointcloud.
-  lib/pc_matrix.c could be split into at least 4 files (linear algebra, point transformations, box transformations and conversions, frustum transformations and conversions)

What is your opinion, @elemoine, @ldgeo, @vpicavet and @pblottiere ?

**Comment by elemoine** [comment]
> pgsql/sql/pointcloud.sql additions may go to a new pgsql/sql/li3ds.sql file (as well as the expected files)

That makes sense. I'll do that.

> pgsql/pointcloud.sql.in additions could go to a new pgsql/li3ds.sql.in file (does it require a new extension ?)

https://www.postgresql.org/docs/9.6/static/extend-extensions.html indicates that an extension may have multiple SQL files. But I don't know how to set it up in practise.

> pc__li3ds files and functions could be renamed to li3ds_* or li_*, at least for most of them that are not depending on pgpointcloud.

You mean you'd use `li3ds` as a prefix rather than a suffix for the file names? So `pc_patch_li3ds` would be `li3ds_pc_patch`? And you would also prefix all the function names in these files?

(I would not use `li` as a prefix/suffix, because I think it's not explicit enough.)


> lib/pc_matrix.c could be split into at least 4 files (linear algebra, point transformations, box transformations and conversions, frustum transformations and conversions)

Yep, that belongs to a separate PR :)

**Comment by mbredif** [comment]
> https://www.postgresql.org/docs/9.6/static/extend-extensions.html indicates that an extension may have multiple SQL files. But I don't know how to set it up in practise.

Neither do I.

> You mean you'd use li3ds as a prefix rather than a suffix for the file names? So pc_patch_li3ds would be li3ds_pc_patch? And you would also prefix all the function names in these files?

Yes that would make it clearer what is intended to be merged upstream and what is li3ds-specific (and could ultimately go to a separate extension).
For functions and files that do not depend on pgpointcloud, we could use li3ds_*, such as `li3ds_frustum` or `li_frustum` and similarly `LIFRUSTUM` instead of `PCFRUSTUM`. For files/functions/types that depend on both, we could go for one of `li3ds_pc_*`, `li3ds_*` or `li_*`, such as `li3ds_pc_patch`, `li3ds_patch`, `lipc_patch` or `li_patch`.

> Yep, that belongs to a separate PR :)

Good.

**Comment by elemoine** [comment]
> pgsql/sql/pointcloud.sql additions may go to a new pgsql/sql/li3ds.sql file (as well as the expected files)

Done.

**Comment by elemoine** [comment]
> For functions and files that do not depend on pgpointcloud, we could use li3ds_*, such as li3ds_frustum or li_frustum and similarly LIFRUSTUM instead of PCFRUSTUM. For files/functions/types that depend on both, we could go for one of li3ds_pc_*, li3ds_* or li_*, such as li3ds_pc_patch, li3ds_patch, lipc_patch or li_patch.

Let's make it simple and just replace `pc` by `li3ds` in the li3ds-specific files and functions. For example:

* `li3ds_matrix.c`
* `li3ds_patch.c`
* `li3ds_patch_rotate_quaternion`
* `LI3DSFRUSTUM`

I could also accept `li` or `l3` as the prefix name, for a prefix with the same number of chars as `pc`, but I also find it less explicit. Please decide, as I don't have a strong opinion on this.

**Comment by mbredif** [comment]
no strong opinion either, but `li` will be less cumbersome.

**Comment by elemoine** [comment]
> pgsql/pointcloud.sql.in additions could go to a new pgsql/li3ds.sql.in file (does it require a new extension ?)

i found a way for this. See the last commit, which I am going to propose upstream.

**Comment by elemoine** [comment]
> no strong opinion either, but li will be less cumbersome.

Ok.

**Comment by elemoine** [comment]
@mbredif see https://github.com/LI3DS/pointcloud/pull/63/commits/7ef6fecb1836dc04c7a18340065763988efac70f. If you agree with this commit I'll squash it into the first commit.

**Comment by elemoine** [comment]
>> pgsql/pointcloud.sql.in additions could go to a new pgsql/li3ds.sql.in file (does it require a new extension ?)
> 
> i found a way for this. See the last commit, which I am going to propose upstream.

I realize that this only works with autotools, and not with cmake. So I think I'll just remove it for now and put everything in `pointcloud.sql.in` as before (and until we find a way to make this work with cmake as well).

**Comment by mbredif** [comment]
elemoine/pointcloud#2 proposes a fix for the cmake support, including multiple sql.in files, but which rely on a `cat` command, as cmake does not seem to provide any way to append files directly.

**Comment by elemoine** [comment]
@mbredif, take a look at https://github.com/LI3DS/pointcloud/pull/63/commits/c0539150ed2567dd9f6477748c20a4f118f86040 (this PR's first commit). This is what I want to propose upstream.

File: pgsql/Makefile [hunk 2]
```diff
@@ -49,8 +49,8 @@ $(EXTENSION).control: $(EXTENSION).control.in Makefile
 	$(SED) -e 's/@POINTCLOUD_VERSION@/$(EXTVERSION)/' \
          -e 's/@POINTCLOUD_VERSION_MAJOR@/$(EXTVERSION_MAJOR)/' $< > $@
 
-$(EXTENSION)--$(EXTVERSION).sql: $(EXTENSION).sql.in Makefile
-	$(SED) -e 's/@POINTCLOUD_VERSION@/$(EXTVERSION)/' $< > $@
+$(EXTENSION)--$(EXTVERSION).sql: $(EXTENSION).sql.in *.sql.in Makefile
```
Comment (mbredif) [inline pgsql/Makefile]: `$(EXTENSION).sql.in` is redundant with `*.sql.in`.

File: pgsql/CMakeLists.txt [hunk 3]
```diff
@@ -16,8 +16,25 @@ set ( PC_INSTALL_EXENSIONS
...
   )
 
+set ( PC_SQLS
+  pointcloud.sql.in
+  )
+
+# create a "cat" function, used for concatenating the sql.in files
+# in PC_SQLS into one sql.in file
+function(cat IN_FILE OUT_FILE)
+  file(READ ${IN_FILE} CONTENTS)
```
Comment (mbredif) [inline pgsql/CMakeLists.txt]: should `${IN_FILE}` and `${OUT_FILE}` be quoted `"${IN_FILE}"` and `"${OUT_FILE}"` ?

**Review by mbredif — COMMENTED** [review]
That looks good. However, I am not sure an upstream PR will make it through, considering there is only a single `sql.in`  file there.

File: pgsql/CMakeLists.txt (hunk 3 above)
Comment (elemoine) [reply pgsql/CMakeLists.txt]: I think it doesn't matter.

File: pgsql/Makefile (hunk 2 above)
Comment (elemoine) [reply pgsql/Makefile]: No, because I want pointcloud.sql.in to always be first in the output.

File: pgsql/li_editor.c [hunk 4]
```diff
@@ -1,33 +1,34 @@
 /***********************************************************************
-* pc_editor.c
+* pc_editor_li3ds.c
```
Comment (mbredif) [inline pgsql/li_editor.c]: That should be li_editor.c

File: pgsql/li_inout.c [hunk 5]
```diff
@@ -0,0 +1,39 @@
+/***********************************************************************
+* pc_inout_li3ds.c
```
Comment (mbredif) [inline pgsql/li_inout.c]: To be renamed

File: pgsql/li_editor.c (hunk 4 above)
Comment (elemoine) [reply pgsql/li_editor.c]: Done. Thanks.

File: pgsql/li_inout.c (hunk 5 above)
Comment (elemoine) [reply pgsql/li_inout.c]: Done.

**Comment by mbredif** [comment]
Good, this can be merged now, thanks @elemoine !

**Comment by elemoine** [comment]
Before merging I'd like to see if the first commit can be merged upstream.

**Comment by elemoine** [comment]
@mbredif patch updated, as discussed. Thanks.

**Review by mbredif — COMMENTED** [review]
That looks good to me. Could you rebase it (you will have conflicts with the new pc_schema_from_xml api) so that I can try it?

**Comment by elemoine** [comment]
I'm going to open a PR for syncing our master branch from upstream master. And this is merged I'll rebase the reorg branch onto master.

**Comment by elemoine** [comment]
@mbredif This is now rebased on latest master.

**Comment by mbredif** [comment]
Looks good ! This can be merged. Thanks @elemoine !
